glommio/src/error.rs

```diff
@@ -24,66 +26,123 @@ pub enum ResourceType {
-    #[error("Channel error: {0}")]
-    ChannelError(#[from] channels::ChannelError<T>),
+    IoError(#[from] io::Error),
 
     /// Queue could not be found error variant
-    #[error("Queue not found error: {0}")]
-    QueueNotFoundError(#[from] executor::QueueNotFoundError),
+    #[error("Queue #{index} not found")]
+    QueueNotFoundError {
+        /// index of the queue in question
+        index: usize,
+    },
 
     /// Queue still active error variant
-    #[error("Queue still active error: {0}")]
-    QueueStillActiveError(#[from] executor::QueueStillActiveError),
-
-    /// try_lock error variant
-    #[error("Try lock error: {0}")]
-    TryLockError(#[from] sync::TryLockError),
+    #[error("Queue #{index} is still active")]
+    QueueStillActiveError {
+        /// index of the queue
+        index: usize,
+    },
 
-    /// Error variant that indicates that the semaphore is closed
-    #[error("Semaphore closed")]
-    SemaphoreClosed,
+    /// The resource in question is closed. Generic because the channel
+    /// variant needs to return the actual item sent into the channel.
+    #[error("{0} is closed")]
+    Closed(ResourceType<T>),
 
     /// Error encapsulating the `WouldBlock` error for types that don't have
     /// errors originating in the standard library. Glommio also has
     /// nonblocking types that need to indicate if they are blocking or not.
     /// This type allows for signaling when a function would otherwise block
     /// for a specific `ResourceType`.
-    #[error("Resource would block. {0}")]
-    WouldBlock(ResourceType),
+    #[error("{0} would block")]
+    WouldBlock(ResourceType<T>),
+}
+
+#[doc(hidden)]
+/// This `Debug` implementation is required, otherwise we'd be required to
+/// place a bound on the generic `T` in GlommioError. This causes the `Debug`
+/// constraint to be forced onto users of the type, and it's an annoying burden
+/// on the type. This gets around that.
+impl<T> Debug for GlommioError<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            GlommioError::IoError(err) => f.write_fmt(format_args!("{:?}", err)),
+            GlommioError::QueueNotFoundError { index }
+            | GlommioError::QueueStillActiveError { index } => {
+                f.write_fmt(format_args!("Queue {{ index: {} }}", index))
+            }
+            GlommioError::Closed(resource) | GlommioError::WouldBlock(resource) => match resource {
+                ResourceType::Semaphore {
+                    requested,
+                    available,
+                } => f.write_fmt(format_args!(
+                    "Semaphore {{ requested {}, available: {} }}",
+                    requested, available
+                )),
+                ResourceType::RwLock => f.write_str("RwLock {{ .. }}"),
+                ResourceType::Channel(_) => f.write_str("Channel {{ .. }}"),
+            },
+        }
+    }
+}
+
+/// TODO: finish
```

Comment:
The ones you added are coverage enough.
We should just add a catch-all statement after that and convert anything else to `Other`